WKD public keys - for now not integrated

ekievsky opened the pull request:
This PR implements logic for retrieving wkd or attester public keys;

issue #296

----------------------------------

**Tests**:
- Tests updated for WKDURLsConstructor
- Can't add tests for WKDURLsApi and AttesterApi because there is no networking entity that injects into these classes and could be spied. 
P.S. It would be great to refactor Networking call `URLSession.shared.call(_ urlRequest: URLRequest, tolerateStatus: [Int]? = nil) -> Promise<HttpRes>` for separate networking entity to test classes that make API calls.

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [x] addresses the issue it closes (if any)
- [x] code is readable and understandable
- [x] is accompanied with tests, or tests are not needed
- [x] is free of vulnerabilities
- [x] is documented clearly and usefully, or doesn't need documentation

tomholub commented on FlowCrypt/Functionality/Services/WKDURLsApi.swift:
```diff
@@ -0,0 +1,129 @@
+        self.core = core
+    }
+
+    func lookupEmail(_ email: String) -> Promise<[String]> {
```
Here we are getting the raw bytes, then parsing them, and then armoring them and returning armored result.

Somewhere down the line another code will de-armor, parse the result to actually use the keys.

Best would be to change the signature of returns of both AttesterApi and this class have a public signature only `lookupEmail(_ email: String) -> Promise<[KeyDetails]>` - just one method

tomholub commented on FlowCrypt/Functionality/WKDURLs/PubLookup.swift:
```diff
@@ -0,0 +1,54 @@
+        self.core = core
+    }
+
+    func lookupEmail(_ email: String) -> Promise<[String]> {
```
And also here - return `Promise<[KeyDetails]>`

tomholub reviewed (commented):
Looks good - see comments. Where is the timeout handled for WKD lookup? There should be a timeout of 4 seconds (Attester could have a timeout of 10 seconds if not already present - please check).

tomholub commented:
@ekievsky please let me know when ready. I understood this is the stage two of #296 - looking forward to get this merged and move onto stage three. Thanks!

ekievsky commented:
Finishing with timeouts, will be done today.

tomholub commented:
> Can't add tests for WKDURLsApi and AttesterApi because there is no networking entity that injects into these classes and could be spied.
> P.S. It would be great to refactor Networking call URLSession.shared.call(_ urlRequest: URLRequest, tolerateStatus: [Int]? = nil) -> Promise<HttpRes> for separate networking entity to test classes that make API calls.
> 

Please update these as appropriate to make them more testable.

ekievsky commented:
@tomholub I've updated PR, resolved your comments and added timeout.

According to networking, I am not sure that it's a good idea to complete it in scope of this task. Because we need to create a separate layer for networking and start using this. I would discuss the implementation with @Kharchevskyi to keep project's code in one design. Please let me know your thought, guys.

tomholub commented:
Ok. I'll have a look

tomholub commented on FlowCrypt/Functionality/Services/WKDURLsApi.swift:
```diff
@@ -37,9 +40,10 @@ class WKDURLsApi: WKDURLsApiType {
             resolve(pubKeys)
         }
+        .timeout(Constants.lookupEmailRequestTimeout)
+        .recoverFromTimeOut(result: [])
```
This is correct for WKD - because the recipient may or may not have a WKD server set up.

However, timeout is not the only failure mode here. If their server is not reachable, that should also result in `[]` (when servers are not set up, sometimes it results in a timeout and sometimes it results in "server not reachable" error right away - need to handle both).

ekievsky replied:
it will return [] if any error occurs with request.

tomholub replied:
Yes and no. 
 - If any error occurred when getting the `/policy` file, then that's a `[]`
 - However, if you are already past the policy stage, and now querying the URL of the public key, then only `HTTP 404` or timeout results in `[]`. Other errors should be escalated as usual.

I'll make an another issue for this.

tomholub reviewed (commented):
Almost there - see comments below.

tomholub commented:
Also please resolve the conflicts. Thanks!

ekievsky commented on FlowCrypt/Functionality/Services/AttesterApi.swift:
```diff
@@ -45,6 +50,8 @@ extension AttesterApi {
             throw AppErr.unexpected("Status \(res.status) when looking up pubkey for \(email)")
         }
+        .timeout(Constants.lookupEmailRequestTimeout)
+        .recoverFromTimeOut(result: PubkeySearchResult(email: email, armored: nil))
```
done, removed recovering from timeout, so it will throw the timeout error in Promise and client that uses this class will be able to catch that error.

tomholub commented on FlowCrypt/Functionality/WKDURLs/PubLookup.swift:
```diff
@@ -0,0 +1,53 @@
+            if !wkdResult.isEmpty {
+                resolve(wkdResult)
+            }
+            let attesterResult = try awaitPromise(self.attesterApi.lookupEmail(email: email))
+            guard let armoredData = attesterResult.armored,
+                  let attesterKeys = try? self.core.parseKeys(armoredOrBinary: armoredData),
+                  !attesterKeys.keyDetails.isEmpty else {
+                resolve([])
+                return
+            }
```
Here I recommend to change attesterApi.lookupEmail result to return `[KeyDetails]` too for consistency (in another PR)
